0.preprocess-sites/8.visualize-plate-layout.py

```diff
@@ -0,0 +1,392 @@
+output_figuresdir = pathlib.Path(summ_cells_figures_basedir, batch)
+cell_category_order = summ_cell_args["cell_category_order"]
+
+correlation_threshold = summ_plate_args["correlation_threshold"]
+painting_image_names = summ_plate_args["painting_image_names"]
+barcoding_cycles = summ_plate_args["barcoding_cycles"]
+barcoding_prefix = summ_plate_args["barcoding_prefix"]
+
+cell_count_file = pathlib.Path(
+    summ_cells_results_basedir, batch, "cells", "cell_count.tsv"
+)
+cell_count_df = pd.read_csv(cell_count_file, sep="\t")
+
+figures_output = pathlib.Path(summ_cells_figures_basedir, batch)
+os.makedirs(figures_output, exist_ok=True)
+results_output = pathlib.Path(summ_cells_results_basedir, batch)
+os.makedirs(results_output, exist_ok=True)
+
+# creates x, y coordinates for plotting per-plate views.
+# assumes image numbering starts in upper left corner and proceeds down
+x_seq = 1
+y_seq = 1
+build_seq = []
+final_order = []
+for i in range(1, ((acquisition_side * acquisition_side) + 1)):
+    if len(final_order) == (acquisition_side * acquisition_side):
+        break
+    else:
+        build_seq.append([x_seq, y_seq])
+        if len(build_seq) == acquisition_side:
+            x_seq = x_seq + 1
+        if y_seq == acquisition_side:
+            y_seq = 0
+            build_seq = build_seq[::-1]
+            final_order.extend(build_seq)
+            build_seq = []
+        y_seq = y_seq + 1
+sites_list = [*range(1, (acquisition_side * acquisition_side) + 1)]
+
+# uses sites_list in case there are fewer analyzed sites than acquired sites
+loc_df = pd.DataFrame(final_order).rename(columns={0: "x_loc", 1: "y_loc"})
+loc_df["Site"] = sites_list
+
+# create total_cell_count
+cell_count_bysite_df = (
+    cell_count_df.groupby("site")["cell_count"]
+    .sum()
+    .reset_index()
+    .rename(columns={"cell_count": "total_cell_count"})
+)
+# add total_cell_count to cell_count_df
+cell_count_df = cell_count_df.merge(cell_count_bysite_df, on="site")
+# add in x, y coordinates for plotting
+cell_count_df = cell_count_df.merge(loc_df, left_on="Site", right_on="Site")
+
+# Plot total number of cells per well
+cell_count_totalcells_df = (
+    cell_count_df.groupby(["x_loc", "y_loc", "Well", "Site"])["total_cell_count"]
+    .mean()
+    .reset_index()
+)
+os.makedirs(output_figuresdir, exist_ok=True)
+
+by_well_gg = (
+    gg.ggplot(cell_count_totalcells_df, gg.aes(x="x_loc", y="y_loc"))
+    + gg.geom_point(gg.aes(fill="total_cell_count"), size=10)
+    + gg.geom_text(gg.aes(label="Site"), color="lightgrey")
+    + gg.facet_wrap("~Well")
+    + gg.coord_fixed()
+    + gg.theme_bw()
+    + gg.theme(
+        axis_text=gg.element_blank(),
+        axis_title=gg.element_blank(),
+        strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
+    )
+    + gg.scale_fill_cmap(name="magma")
+)
+output_file = pathlib.Path(output_figuresdir, "plate_layout_cells_count_per_well.png")
+by_well_gg.save(output_file, dpi=300, verbose=False)
```

Comment:
maybe beyond scope of this PR (we can leave for a future refactor, if you think it's best) is to pull out this plotting logic into a function and add to an import. We plot a whole bunch of very similar plots in this script with the only thing changingas `fill="total_cell_count"` IIUC

Reply:
Can you explain what we would gain by pulling the similar plots out into a function?

Reply:
two benefits:

* Every time we plot something, we take ~16 lines of code. We can reduce this to 1 line. This follows [DRY principles](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
* One can imagine updating these plots in the future - if we had a function, all we would need to do is update the function.

There is definitely a balance though. If the figures are slightly different across different `fill` arguments, then a function might be too clumsy.